### PR TITLE
fixed the poolSubpage.maxNumElems == 1, free subpage bug

### LINKED ISSUE #10896: When subpage.elemSize == 8192, the free subpage operation is not work
### Actual behavior
allocate N * 8192 subpage, free part of them, but subpage still esixt in two-way linked list 

### Minimal yet complete reproducer code (or URL to code)
```
io.netty.buffer.PoolSubpage#free {
    if (numAvail ++ == 0) { // numAvail always equals 0 when free subpage and elemSize == 8192
        addToPool(head);
        return true;
    }
}
```
### Netty version
4.1.52
### JVM version (e.g. `java -version`)
1.8

### PR DESCRIPTION
**Pull request by ZzxyNn**
Motivation:

when customer need large of 'byteBuf.capacity' in [7168, 8192], the size of 'chunk.subpages' may be inflated when large of byteBuf be released, not consistent with other 'byteBuf.capacity'

Modification:

when maxNumElems == 1 need consider remove from pool

Result:

Fixes #10896 .

### CODE REVIEW

File: buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java [hunk 1]
```diff
@@ -1,10 +1,31 @@
...
+
 package io.netty.buffer;
 
 import org.junit.platform.commons.util.ReflectionUtils;
 
+import java.lang.reflect.Field;
+
 import static org.junit.Assert.assertEquals;
 
 public class PoolSubpageReleaseTest {
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:25]: unit tests should use junit

File: buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java [hunk 2]
```diff
@@ -1,10 +1,31 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:2]: 2020

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        return num;
+    }
+
+    private static PoolSubpage[] byReflectionGetSubpages(ByteBuf one) {
+        Object chunk;
+        try {
+            chunk = ReflectionUtils.tryToReadFieldValue(PooledByteBuf.class.getDeclaredField("chunk"), one).get();
+            Field subpagesField = PoolChunk.class.getDeclaredField("subpages");
+            return (PoolSubpage[]) ReflectionUtils.tryToReadFieldValue(subpagesField, chunk).get();
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:93]: Please expose field as package private so it can be accessed here without reflection

File: buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        //first release the 'tows', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
+        // ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
+        for (int i = 0; i < fixLength; i++) {
+            fixByteBuf[i].release();
+        }
+        // release the 'ones'
+        for (int i = 0; i < oneLength; i++) {
+            ones[i].release();
+            // System.out.println("subpages size: " + countSubpages(subpages));
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:71]: Remove

File: buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    private void run(int elemSize) {
+        int oneLength = 128, fixLength = 256;
+        ByteBuf[] ones = new ByteBuf[oneLength];
+        ByteBuf[] fixByteBuf = new ByteBuf[fixLength];
+
+        //allocate
+        for (int i = 0; i < fixLength; i++) {
+            fixByteBuf[i] = ByteBufAllocator.DEFAULT.buffer(elemSize, elemSize);
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:50]: You should directly instance a PooledByteBufAllocator here... or even merge this into PooledByteBufAllocatorTest

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java (hunk 3 above)
Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:93]: which field expose package private,  subpages ?

Comment (normanmaurer) [reply buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:93]: Yes

Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:93]: for a unit test?

Comment (normanmaurer) [reply buffer/src/test/java/io/netty/buffer/PoolSubpageReleaseTest.java:93]: yes... Reflection should only be used as last resort... just make it package-private and add a comment that it is only package-private for testing.

**Comment by ZzxyNn** [comment]
@normanmaurer 
what’s else problem?

**Comment by hyperxpro** [comment]
```
Run ./.github/scripts/check_leak.sh build-leak.output
Leak detected, please inspect build log
Error: Process completed with exit code 1.
```

Fix `ByteBuf` leaks.

**Comment by ZzxyNn** [comment]
> ```
> Run ./.github/scripts/check_leak.sh build-leak.output
> Leak detected, please inspect build log
> Error: Process completed with exit code 1.
> ```
> 
> Fix `ByteBuf` leaks.

i tried to remove the unit tests i added, then the Leak detected passes, but in the unit tests i added , all ByteBuf calls release.

**Comment by mostroverkhov** [comment]
@ZzxyNn 
There is `ClassCastException` in test in commit [675c4b](https://github.com/netty/netty/runs/1621650494), It is thrown before buffers release statements are reached
```
2020-12-29T12:25:52.3497652Z [ERROR] testNormalPoolSubpageRelease  Time elapsed: 0 s  <<< ERROR!
2020-12-29T12:25:52.3499472Z java.lang.ClassCastException: io.netty.buffer.AdvancedLeakAwareByteBuf cannot be cast to io.netty.buffer.PooledUnsafeHeapByteBuf
2020-12-29T12:25:52.3501741Z 	at io.netty.buffer.PooledByteBufAllocatorTest.run(PooledByteBufAllocatorTest.java:689)
2020-12-29T12:25:52.3504287Z 	at io.netty.buffer.PooledByteBufAllocatorTest.testNormalPoolSubpageRelease(PooledByteBufAllocatorTest.java:676)
```

**Comment by ZzxyNn** [comment]
ok，its all checks passed

**Comment by normanmaurer** [comment]
@ZzxyNn so I am back at work and had another look.

WDYT about applying this change on top of yours:

```
diff --git a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
index b3879ff433..4d7a5ebb9e 100644
--- a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
+++ b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
@@ -632,8 +632,8 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
     }
 
     @SuppressWarnings("unchecked")
-    private static PooledByteBuf<ByteBuffer> unwrapIfNeeded(ByteBuf buf) {
-        return (PooledByteBuf<ByteBuffer>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
+    private static <T> PooledByteBuf<T> unwrapIfNeeded(ByteBuf buf) {
+        return (PooledByteBuf<T>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
     }
 
     @Test
@@ -667,27 +667,17 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
         // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
         // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
         int elemSize = 32;
-        run(elemSize, false);
-    }
-
-    private void run(int elemSize, boolean preferDirect) {
-        int oneLength = 16, fixedLength = 256;
+        int oneLength = 16;
+        int fixedLength = 256;
         ByteBuf[] ones = new ByteBuf[oneLength];
         ByteBuf[] fixedByteBuf = new ByteBuf[fixedLength];
-        PooledByteBufAllocator allocator = newAllocator(preferDirect);
+        final PooledByteBufAllocator allocator = newAllocator(false);
 
-        //allocate
         for (int i = 0; i < fixedLength; i++) {
             fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
         }
 
-        if (!PooledByteBuf.class.isAssignableFrom(fixedByteBuf[0].getClass())) {
-            for (int i = 0; i < fixedLength; i++) {
-                fixedByteBuf[i].release();
-            }
-            return;
-        }
-        PoolSubpage[] subpages = ((PooledByteBuf) fixedByteBuf[0]).chunk.subpages;
+        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
 
         // fixSubpagesCount ==  when ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
         // the chunk.poolSubpages.size
@@ -697,7 +687,7 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
             ones[i] = allocator.heapBuffer(elemSize, elemSize);
         }
 
-        //first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
+        // first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
         // ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
         for (int i = 0; i < fixedLength; i++) {
             fixedByteBuf[i].release();
@@ -711,10 +701,10 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
         assertEquals(countSubpages(subpages), fixedSubpagesCount + 1);
     }
 
-    private static int countSubpages(PoolSubpage<Object>[] subpages) {
+    private static int countSubpages(PoolSubpage<?>[] subpages) {
         int num = 0;
-        for (final PoolSubpage<Object> subpage : subpages) {
-            if (null != subpage) {
+        for (final PoolSubpage<?> subpage : subpages) {
+            if (subpage != null) {
                 num++;
             }
         }

```

**Comment by ZzxyNn** [comment]
> @ZzxyNn so I am back at work and had another look.
> 
> WDYT about applying this change on top of yours:
> 
> ```
> diff --git a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
> index b3879ff433..4d7a5ebb9e 100644
> --- a/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
> +++ b/buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java
> @@ -632,8 +632,8 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>      }
>  
>      @SuppressWarnings("unchecked")
> -    private static PooledByteBuf<ByteBuffer> unwrapIfNeeded(ByteBuf buf) {
> -        return (PooledByteBuf<ByteBuffer>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
> +    private static <T> PooledByteBuf<T> unwrapIfNeeded(ByteBuf buf) {
> +        return (PooledByteBuf<T>) (buf instanceof PooledByteBuf ? buf : buf.unwrap());
>      }
>  
>      @Test
> @@ -667,27 +667,17 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>          // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
>          // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
>          int elemSize = 32;
> -        run(elemSize, false);
> -    }
> -
> -    private void run(int elemSize, boolean preferDirect) {
> -        int oneLength = 16, fixedLength = 256;
> +        int oneLength = 16;
> +        int fixedLength = 256;
>          ByteBuf[] ones = new ByteBuf[oneLength];
>          ByteBuf[] fixedByteBuf = new ByteBuf[fixedLength];
> -        PooledByteBufAllocator allocator = newAllocator(preferDirect);
> +        final PooledByteBufAllocator allocator = newAllocator(false);
>  
> -        //allocate
>          for (int i = 0; i < fixedLength; i++) {
>              fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
>          }
>  
> -        if (!PooledByteBuf.class.isAssignableFrom(fixedByteBuf[0].getClass())) {
> -            for (int i = 0; i < fixedLength; i++) {
> -                fixedByteBuf[i].release();
> -            }
> -            return;
> -        }
> -        PoolSubpage[] subpages = ((PooledByteBuf) fixedByteBuf[0]).chunk.subpages;
> +        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
>  
>          // fixSubpagesCount ==  when ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
>          // the chunk.poolSubpages.size
> @@ -697,7 +687,7 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>              ones[i] = allocator.heapBuffer(elemSize, elemSize);
>          }
>  
> -        //first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
> +        // first release the 'fixByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
>          // ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
>          for (int i = 0; i < fixedLength; i++) {
>              fixedByteBuf[i].release();
> @@ -711,10 +701,10 @@ public class PooledByteBufAllocatorTest extends AbstractByteBufAllocatorTest<Poo
>          assertEquals(countSubpages(subpages), fixedSubpagesCount + 1);
>      }
>  
> -    private static int countSubpages(PoolSubpage<Object>[] subpages) {
> +    private static int countSubpages(PoolSubpage<?>[] subpages) {
>          int num = 0;
> -        for (final PoolSubpage<Object> subpage : subpages) {
> -            if (null != subpage) {
> +        for (final PoolSubpage<?> subpage : subpages) {
> +            if (subpage != null) {
>                  num++;
>              }
>          }
> ```

emmm, clearly and simply

**Comment by normanmaurer** [comment]
Please apply then

**Comment by ZzxyNn** [comment]
> Please apply then

it's done

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 6]
```diff
@@ -661,4 +661,53 @@ public void testCacheWorksForNormalAllocations() {
...
+
+        // fixSubpagesCount ==  when ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
+        // the chunk.poolSubpages.size
+        int fixedSubpagesCount = countSubpages(subpages);
+
+        for (int i = 0; i < oneLength; i++) {
+            ones[i] = allocator.heapBuffer(elemSize, elemSize);
+        }
+
+        // first release the 'fixedByteBuf', in oder to filled of PoolThreadCache.MemoryRegionCache.queue
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:690]:
```suggestion
        // first release the 'fixedByteBuf', in order to fill the PoolThreadCache.MemoryRegionCache.queue
```

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java [hunk 7]
```diff
@@ -115,7 +115,9 @@ boolean free(PoolSubpage<T> head, int bitmapIdx) {
 
         if (numAvail ++ == 0) {
             addToPool(head);
-            return true;
+            if (maxNumElems > 1) {
+                return true;
+            }
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/PoolSubpage.java:120]: The test you added passes without this change. So this is not really being tested.

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 8]
```diff
@@ -661,4 +661,53 @@ public void testCacheWorksForNormalAllocations() {
...
+        int fixedLength = 256;
+        ByteBuf[] ones = new ByteBuf[oneLength];
+        ByteBuf[] fixedByteBuf = new ByteBuf[fixedLength];
+        final PooledByteBufAllocator allocator = newAllocator(false);
+
+        for (int i = 0; i < fixedLength; i++) {
+            fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
+        }
+
+        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:680]: Try to write a test that verifies on the externally observable behaviour of the allocator, rather than asserting on internal state changes.

Tests that rely on implementation internals tend to be more brittle, and easier to break on unrelated changes.

**Review by chrisvest — CHANGES_REQUESTED** [review]
The test seems to be inadequate.

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java (hunk 8 above)
Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:680]: the externally observable behavior of the allocator is when 'maxNumElems == 1' then release subpage operate not work, have no idea how to asserting that @chrisvest

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java (hunk 7 above)
Comment (ZzxyNn) [reply buffer/src/main/java/io/netty/buffer/PoolSubpage.java:120]: Has been adjusted.

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java (hunk 6 above)
Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java]: Has been adjusted.

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 9]
```diff
@@ -666,48 +666,25 @@ public void testCacheWorksForNormalAllocations() {
...
+        PooledByteBuf<Object> pooledByteBuf = unwrapIfNeeded(byteBufs[0]);
+        // disable PoolThreadCache
+        pooledByteBuf.cache = null;
+        PoolChunk<Object> chunk = pooledByteBuf.chunk;
+
+        int beforeFreeBytes = chunk.freeBytes();
+        for (int i = 0; i < length; i++) {
+            byteBufs[i].release();
         }
+        int AfterFreeBytes = chunk.freeBytes();
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:686]: nit: s/After.../after../

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 10]
```diff
@@ -666,48 +666,25 @@ public void testCacheWorksForNormalAllocations() {
...
-
-        // fixSubpagesCount ==  when ( PoolThreadCache.MemoryRegionCache.queue.offer(entry) == false )
-        // the chunk.poolSubpages.size
-        int fixedSubpagesCount = countSubpages(subpages);
-
-        for (int i = 0; i < oneLength; i++) {
-            ones[i] = allocator.heapBuffer(elemSize, elemSize);
+        PooledByteBuf<Object> pooledByteBuf = unwrapIfNeeded(byteBufs[0]);
+        // disable PoolThreadCache
+        pooledByteBuf.cache = null;
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:679]: don't do this... you can construct the `PooledByteBufAllocator` with caches disabled.

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 11]
```diff
@@ -666,48 +666,25 @@ public void testCacheWorksForNormalAllocations() {
...
+        ByteBuf[] byteBufs = new ByteBuf[length];
         final PooledByteBufAllocator allocator = newAllocator(false);
 
-        for (int i = 0; i < fixedLength; i++) {
-            fixedByteBuf[i] = allocator.heapBuffer(elemSize, elemSize);
+        for (int i = 0; i < length; i++) {
+            byteBufs[i] = allocator.heapBuffer(elemSize, elemSize);
         }
-
-        PoolSubpage<?>[] subpages = unwrapIfNeeded(fixedByteBuf[0]).chunk.subpages;
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java]: I think this means that `subpages` can be private again ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java (hunk 10 above)
Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java]: OK, please recheck.

Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java]: @normanmaurer How's it going?

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 12]
```diff
@@ -661,4 +665,27 @@ public void testCacheWorksForNormalAllocations() {
         assertEquals(3, allocator.metric().directArenas().get(0).numNormalAllocations());
         buffer.release();
     }
+
+    @Test
+    public void testNormalPoolSubpageRelease() {
+        // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
+        // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
+        int elemSize = 1024;
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:673]: If you make `elemSize = 8192` then this test will fail when your `PoolSubpage` changes are commented out.

```suggestion
        int elemSize = 8192;
```

**Review by chrisvest — CHANGES_REQUESTED** [review]
Just one thing

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java (hunk 12 above)
Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:673]: yes, is there a problem? @chrisvest

Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:673]: If my change is commented and the unit test fails, doesn't that mean the change is correct

Comment (chrisvest) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:673]: Yes. My point is: when your `PoolSubpage` changes are commented out, the test does _not_ fail when the `elemSize` is 1024, but it _does_ fail when `elemSize` is 8192, which is why I'm suggesting making this change.

With your `PoolSubpage` changes in place, the test of course pass for either `elemSize`.

Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:673]: oh, I see what you mean. I did this before, but you said that my changes were not covered by the unit tests

**Comment by ZzxyNn** [comment]
@chrisvest 
`Error: Cannot retrieve repository metadata (repomd.xml) for repository: base. Please verify its path and try again.`

What's going on here

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java [hunk 13]
```diff
@@ -115,7 +115,9 @@ boolean free(PoolSubpage<T> head, int bitmapIdx) {
 
         if (numAvail ++ == 0) {
             addToPool(head);
-            return true;
+            if (maxNumElems > 1) {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolSubpage.java:118]: @ZzxyNn can you just please add a comment here any explain why we return true only if `maxNumElems > 1` ? This will make it easier to understand this in the future.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java [hunk 14]
```diff
@@ -197,6 +198,13 @@ public PooledByteBufAllocator(boolean preferDirect) {
         this(preferDirect, DEFAULT_NUM_HEAP_ARENA, DEFAULT_NUM_DIRECT_ARENA, DEFAULT_PAGE_SIZE, DEFAULT_MAX_ORDER);
     }
 
+    @SuppressWarnings("deprecation")
+    public PooledByteBufAllocator(boolean preferDirect, boolean useCacheForAllThreads) {
+        this(preferDirect, DEFAULT_NUM_HEAP_ARENA, DEFAULT_NUM_DIRECT_ARENA, DEFAULT_PAGE_SIZE, DEFAULT_MAX_ORDER, 0,
+                DEFAULT_SMALL_CACHE_SIZE, DEFAULT_NORMAL_CACHE_SIZE, useCacheForAllThreads,
+                DEFAULT_DIRECT_MEMORY_CACHE_ALIGNMENT);
+    }
+
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:207]: Basically all the changes in this class should be reverted

File: buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java [hunk 15]
```diff
@@ -661,4 +665,27 @@ public void testCacheWorksForNormalAllocations() {
...
     }
+
+    @Test
+    public void testNormalPoolSubpageRelease() {
+        // 16 < elemSize <= 7168 or 8192 < elemSize <= 28672, 1 < subpage.maxNumElems <= 256
+        // 7168 <= elemSize <= 8192, subpage.maxNumElems == 1
+        int elemSize = 8192;
+        int length = 1024;
+        ByteBuf[] byteBufs = new ByteBuf[length];
+        final PooledByteBufAllocator allocator = newAllocator(false, false);
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java:676]: The above

Comment (ZzxyNn) [reply buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java]: please review..

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java [hunk 16]
```diff
@@ -115,7 +115,10 @@ boolean free(PoolSubpage<T> head, int bitmapIdx) {
 
         if (numAvail ++ == 0) {
             addToPool(head);
-            return true;
+            // when maxNumElems == 1, cannot return directly, also need to go through the following judgment
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolSubpage.java:118]: Please reword... This comment doesn't really explain what this is all about.

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/PoolSubpage.java (hunk 16 above)
Comment (ZzxyNn) [reply buffer/src/main/java/io/netty/buffer/PoolSubpage.java:118]: If `maxNumeLems == 1`, the maximum `numAvail` is also 1, and each free operation of the `PoolSubpage` will return in the branch of `numAvail ++ == 0`, which means the following code is unreachable

how about this?

Comment (ZzxyNn) [reply buffer/src/main/java/io/netty/buffer/PoolSubpage.java:118]: If `maxNumElems == 1`, the maximum `numAvail` is also 1. The free operation of this type of `PoolSubPage` will end here, which means that subsequent code is not reachable, these `PoolSubPages` will not really be recycled, so the `maxNumElems > 1` judgment needs to be added

Comment (normanmaurer) [reply buffer/src/main/java/io/netty/buffer/PoolSubpage.java:118]: sounds good... Please update the comment to something like this^^^

Comment (ZzxyNn) [reply buffer/src/main/java/io/netty/buffer/PoolSubpage.java]: alright.

**Comment by ZzxyNn** [comment]
`transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollSocketShutdownOutputBySelfTest.java:1 | expected null, but was:<true>`

What's going on here?

**Comment by normanmaurer** [comment]
@ZzxyNn flaky test... triggered the CI again

**Comment by ZzxyNn** [comment]
> @ZzxyNn flaky test... triggered the CI again

`Error: Resource not accessible by integration`
em....?

**Comment by normanmaurer** [comment]
This will be fixed by https://github.com/netty/netty/pull/10968... I will re-trigger the tests when this is pulled in.

**Comment by ZzxyNn** [comment]
> This will be fixed by #10968... I will re-trigger the tests when this is pulled in.

fine.

**Comment by normanmaurer** [comment]
@ZzxyNn can you rebase so the CI changes are pulled in as well. ?

**Comment by normanmaurer** [comment]
@chrisvest can you have a final look ?

**Review by normanmaurer — APPROVED** [review]

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@ZzxyNn Ok I think this one is ready to go now... Can you please squash the commits and adjust the commit message to explain exactly what this fixes etc ?

**Comment by ZzxyNn** [comment]
> @ZzxyNn Ok I think this one is ready to go now... Can you please squash the commits and adjust the commit message to explain exactly what this fixes etc ?

Oh...I feel like I'm having a problem with my merge.

**Comment by ZzxyNn** [comment]
@normanmaurer em......?

**Comment by normanmaurer** [comment]
@ZzxyNn seems like the commits were not squashed yet and the commit message adjusted... Can you please do this ?

**Comment by normanmaurer** [comment]
See https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

**Comment by ZzxyNn** [comment]
> @ZzxyNn seems like the commits were not squashed yet and the commit message adjusted... Can you please do this ?

little hard for me...

**Comment by ZzxyNn** [comment]
holy ...

**Comment by chrisvest** [comment]
@ZzxyNn Yeah, rebasing can be troublesome when there are merge commits involved. If it's easier for you to make a new branch with the final changes, then just do that.
